ONNX and PyTorch model from RandomForestClassifier have different prediction results

When I try to convert a RandomForestClassifier model to ONNX and Pytorch format, the prediction results from these two models fail to match, here is the example code

```python
import numpy as np
import onnxruntime as ort
import torch
from hummingbird.ml import convert
from onnxconverter_common import FloatTensorType
from onnxmltools import convert_sklearn
from sklearn.datasets import load_breast_cancer
from sklearn.ensemble import RandomForestClassifier

X_bc, y_bc = load_breast_cancer(return_X_y=True)
nrows = 15000
X_bc: np.ndarray = X_bc[0:nrows]
y_bc: np.ndarray = y_bc[0:nrows]

if __name__ == "__main__":
    sklearn_model = RandomForestClassifier(n_estimators=10, max_depth=10)
    sklearn_model.fit(X_bc, y_bc)

    sample_input = torch.rand(100, X_bc.shape[1], dtype=torch.float32)
    sklearn_model_predict = sklearn_model.predict(sample_input.numpy())

    onnx_ml_model = convert_sklearn(
        sklearn_model, initial_types=[("input", FloatTensorType([sample_input.shape[0], sample_input.shape[1]]))],
        target_opset=11
    )
    session = ort.InferenceSession(onnx_ml_model.SerializeToString())
    output_names = [session.get_outputs()[i].name for i in range(len(session.get_outputs()))]
    inputs = {session.get_inputs()[0].name: sample_input.numpy()}
    onnx_ml_model_pred = session.run(output_names, inputs)[0].flatten()

    pt_model = convert(sklearn_model, "torch", X_bc)
    pt_model_pred = pt_model.predict(sample_input)

    np.testing.assert_allclose(onnx_ml_model_pred, pt_model_pred, rtol=1e-5, atol=0)
```

and here is the output

```
AssertionError: 
Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 9 / 100 (9%)
Max absolute difference: 1
Max relative difference: 1.
 x: array([1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,...
 y: array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,...
```

--- a/hummingbird/ml/operator_converters/_gbdt_commons.py
+++ b/hummingbird/ml/operator_converters/_gbdt_commons.py
@@ -42,6 +42,15 @@ def convert_gbdt_classifier_common(
         n_classes -= 1
     if classes is None:
         classes = [i for i in range(n_classes)]
+    # There is a bug in torch < 1.7.0 that causes a mismatch. See Issue #10
+    if n_classes > 2:
+        from distutils.version import LooseVersion
+        import torch
+
+        if LooseVersion(torch.__version__) < LooseVersion("1.7.0"):
+            import warnings
+
+            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass. See issue #10.")
     reorder_trees = True
     if constants.REORDER_TREES in extra_config:
         reorder_trees = extra_config[constants.REORDER_TREES]

--- a/hummingbird/ml/operator_converters/sklearn/decision_tree.py
+++ b/hummingbird/ml/operator_converters/sklearn/decision_tree.py
@@ -36,7 +36,7 @@ def convert_sklearn_random_forest_classifier(operator, device, extra_config):
     classes = operator.raw_operator.classes_.tolist()
 
     # There is a bug in torch < 1.7.0 that causes a mismatch. See Issue #10
-    if classes > 2:
+    if len(classes) > 2:
         from distutils.version import LooseVersion
         import torch
 

--- a/hummingbird/ml/operator_converters/sklearn/decision_tree.py
+++ b/hummingbird/ml/operator_converters/sklearn/decision_tree.py
@@ -35,6 +35,16 @@ def convert_sklearn_random_forest_classifier(operator, device, extra_config):
     n_features = operator.raw_operator.n_features_
     classes = operator.raw_operator.classes_.tolist()
 
+    # There is a bug in torch < 1.7.0 that causes a mismatch. See Issue #10
+    if classes > 2:
+        from distutils.version import LooseVersion
+        import torch
+
+        if LooseVersion(torch.__version__) < LooseVersion("1.7.0"):
+            import warnings
+
+            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass RF. See issue #10.")
+
     # For Sklearn Trees we need to know how many trees are there for normalization.
     extra_config[constants.NUM_TREES] = len(tree_infos)
 

--- a/hummingbird/ml/operator_converters/sklearn/decision_tree.py
+++ b/hummingbird/ml/operator_converters/sklearn/decision_tree.py
@@ -43,7 +43,7 @@ def convert_sklearn_random_forest_classifier(operator, device, extra_config):
         if LooseVersion(torch.__version__) < LooseVersion("1.7.0"):
             import warnings
 
-            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass RF. See issue #10.")
+            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass. See issue #10.")
 
     # For Sklearn Trees we need to know how many trees are there for normalization.
     extra_config[constants.NUM_TREES] = len(tree_infos)

--- a/hummingbird/ml/operator_converters/sklearn/linear.py
+++ b/hummingbird/ml/operator_converters/sklearn/linear.py
@@ -33,7 +33,7 @@ def convert_sklearn_linear_model(operator, device, extra_config):
     supported_loss = {"log", "modified_huber", "squared_hinge"}
     classes = [0] if not hasattr(operator.raw_operator, "classes_") else operator.raw_operator.classes_
     # There is a bug in torch < 1.7.0 that causes a mismatch. See Issue #10
-    if classes > 2:
+    if len(classes) > 2:
         from distutils.version import LooseVersion
         import torch
 

--- a/hummingbird/ml/operator_converters/sklearn/linear.py
+++ b/hummingbird/ml/operator_converters/sklearn/linear.py
@@ -40,7 +40,7 @@ def convert_sklearn_linear_model(operator, device, extra_config):
         if LooseVersion(torch.__version__) < LooseVersion("1.7.0"):
             import warnings
 
-            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass RF. See issue #10.")
+            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass. See issue #10.")
 
     if not all(["int" in str(type(x)) for x in classes]):
         raise RuntimeError(

--- a/hummingbird/ml/operator_converters/sklearn/linear.py
+++ b/hummingbird/ml/operator_converters/sklearn/linear.py
@@ -32,6 +32,15 @@ def convert_sklearn_linear_model(operator, device, extra_config):
 
     supported_loss = {"log", "modified_huber", "squared_hinge"}
     classes = [0] if not hasattr(operator.raw_operator, "classes_") else operator.raw_operator.classes_
+    # There is a bug in torch < 1.7.0 that causes a mismatch. See Issue #10
+    if classes > 2:
+        from distutils.version import LooseVersion
+        import torch
+
+        if LooseVersion(torch.__version__) < LooseVersion("1.7.0"):
+            import warnings
+
+            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass RF. See issue #10.")
 
     if not all(["int" in str(type(x)) for x in classes]):
         raise RuntimeError(

--- a/hummingbird/ml/operator_converters/sklearn/mlp.py
+++ b/hummingbird/ml/operator_converters/sklearn/mlp.py
@@ -31,7 +31,7 @@ def convert_sklearn_mlp_classifier(operator, device, extra_config):
 
     classes = operator.raw_operator.classes_
     # There is a bug in torch < 1.7.0 that causes a mismatch. See Issue #10
-    if classes > 2:
+    if len(classes) > 2:
         from distutils.version import LooseVersion
         import torch
 

--- a/hummingbird/ml/operator_converters/sklearn/mlp.py
+++ b/hummingbird/ml/operator_converters/sklearn/mlp.py
@@ -38,7 +38,7 @@ def convert_sklearn_mlp_classifier(operator, device, extra_config):
         if LooseVersion(torch.__version__) < LooseVersion("1.7.0"):
             import warnings
 
-            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass RF. See issue #10.")
+            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass. See issue #10.")
 
     if not all([type(x) in [int, np.int32, np.int64] for x in classes]):
         raise RuntimeError("Hummingbird supports only integer labels for class labels.")

--- a/hummingbird/ml/operator_converters/sklearn/mlp.py
+++ b/hummingbird/ml/operator_converters/sklearn/mlp.py
@@ -30,6 +30,16 @@ def convert_sklearn_mlp_classifier(operator, device, extra_config):
     assert operator is not None, "Cannot convert None operator"
 
     classes = operator.raw_operator.classes_
+    # There is a bug in torch < 1.7.0 that causes a mismatch. See Issue #10
+    if classes > 2:
+        from distutils.version import LooseVersion
+        import torch
+
+        if LooseVersion(torch.__version__) < LooseVersion("1.7.0"):
+            import warnings
+
+            warnings.warn("torch < 1.7.0 may give a mismatch on multiclass RF. See issue #10.")
+
     if not all([type(x) in [int, np.int32, np.int64] for x in classes]):
         raise RuntimeError("Hummingbird supports only integer labels for class labels.")
 

